homeassistant/components/template/alarm_control_panel.py

```diff
@@ -0,0 +1,316 @@
+        self._arm_away_script = None
+        if arm_away_action is not None:
+            self._arm_away_script = Script(hass, arm_away_action)
+        self._arm_home_script = None
+        if arm_home_action is not None:
+            self._arm_home_script = Script(hass, arm_home_action)
+        self._arm_night_script = None
+        if arm_night_action is not None:
+            self._arm_night_script = Script(hass, arm_night_action)
+
+        self._state = None
+        self._entities = template_entity_ids
+
+        if self._template is not None:
+            self._template.hass = self.hass
+
+    @property
+    def name(self):
+        """Return the display name of this alarm control panel."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def supported_features(self) -> int:
+        """Return the list of supported features."""
+        supported_features = (
+            SUPPORT_ALARM_ARM_NIGHT if (self._arm_night_script is not None) else 0
+        )
+        supported_features = (
+            supported_features | SUPPORT_ALARM_ARM_HOME
+            if (self._arm_home_script is not None)
+            else 0
+        )
+        supported_features = (
+            supported_features | SUPPORT_ALARM_ARM_AWAY
+            if (self._arm_away_script is not None)
+            else 0
+        )
+        return supported_features
+
+    @property
+    def code_format(self):
+        """Return one or more digits/characters."""
+        return FORMAT_NUMBER
+
+    @property
+    def code_arm_required(self):
+        """Whether the code is required for arm actions."""
+        return self._code_arm_required
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+
+        @callback
+        def template_alarm_state_listener(entity, old_state, new_state):
+            """Handle target device state changes."""
+            self.async_schedule_update_ha_state(True)
+
+        @callback
+        def template_alarm_control_panel_startup(event):
+            """Update template on startup."""
+            if self._template is not None:
+                async_track_state_change(
+                    self.hass, self._entities, template_alarm_state_listener
+                )
+
+            self.async_schedule_update_ha_state(True)
+
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, template_alarm_control_panel_startup
+        )
+
+    async def async_alarm_arm_away(self, code=None):
+        """Arm the panel to Away."""
+        optimistic_set = False
+
+        if self._template is None:
+            self._state = STATE_ALARM_ARMED_AWAY
+            optimistic_set = True
+
+        if self._arm_away_script is not None:
+            await self._arm_away_script.async_run(
```

Comment:
All the actuator methods have the same structure. We can make a helper method that does this job and accepts the state and script attribute.

```py
async def _async_alarm_arm(self, state, script=None, code=None):
    """Arm the panel to state."""
    optimistic_set = False

    if self._template is None:
        self._state = state
        optimistic_set = True

    if script is not None:
        await script.async_run(
            {ATTR_CODE: code}, context=self._context
        )
    else:
        _LOGGER.error("No script action defined for %s", state)

    if optimistic_set:
        self.async_schedule_update_ha_state()
```

Reply:
Pah, what's a couple of unnecessary code bytes? ;p. Concur - done.